pkg/controller/release/release_controller.go

```diff
@@ -559,26 +564,26 @@ func reasonForReleaseCondition(err error) string {
 	return "unknown error! tell Shipper devs to classify it"
 }
 
-func (c *Controller) chooseClusters(rel *shipper.Release, force bool) error {
+func (c *Controller) chooseClusters(rel *shipper.Release, force bool) (*shipper.Release, error) {
 	metaKey := controller.MetaKey(rel)
 	if !force && releaseHasClusters(rel) {
-		return shippererrors.NewUnrecoverableError(fmt.Errorf("release %q has already been assigned to clusters", metaKey))
+		return nil, shippererrors.NewUnrecoverableError(fmt.Errorf("release %q has already been assigned to clusters", metaKey))
 	}
 	glog.Infof("Choosing clusters for release %q", metaKey)
 
 	selector := labels.Everything()
 	allClusters, err := c.clusterLister.List(selector)
 	if err != nil {
-		return shippererrors.NewKubeclientListError(
+		return nil, shippererrors.NewKubeclientListError(
 			shipper.SchemeGroupVersion.WithKind("Cluster"),
 			"", selector, err)
 	}
 
 	selectedClusters, err := computeTargetClusters(rel, allClusters)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	setReleaseClusters(rel, selectedClusters)
```

Comment:
`chooseClusters` could return `nil, nil` here to indicate that no change has been applied.